src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SemanticRange.cs

```diff
@@ -2,25 +2,47 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Diagnostics;
 using OmniSharp.Extensions.LanguageServer.Protocol.Models;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
 {
-    internal class SemanticRange
+    internal class SemanticRange : IComparable<SemanticRange>
     {
-        public SemanticRange(int kind, Range range)
+        public SemanticRange(int kind, Range range, int modifier)
         {
             if (range is null)
             {
                 throw new ArgumentNullException(nameof(range));
             }
 
             Kind = kind;
+            Modifier = modifier;
             Range = range;
         }
 
         public Range Range { get; set; }
 
         public int Kind { get; set; }
+
+        public int Modifier { get; set; }
+
+        public int CompareTo(SemanticRange other)
+        {
+            if(other is null)
+            {
+                return 1;
+            }
+
+            if(other.Range is null && Range is null)
+            {
+                return 0;
+            }
+
+            Debug.Assert(Range.Start.CompareTo(other.Range.Start) == 0 || !Range.OverlapsWith(other.Range));
+
+            // Since overlapping SemanticRanges are STRICTLY FORBIDDEN we need only compare the starts 
```

Comment:
We combine multiple languages semantic tokens though. Do we validate this statement elsewhere?

Reply:
Just above this we Debug.Assert that the Ranges do not overlap. I didn't want to slow down the process by checking it in Release since it should be theoretically impossible. As far as I know the multiple language servers thing could only be a confounding factor if either a token was considered to be both a Razor token and a CSharp token or there was a bug in converting/remapping their SemanticTokens to SemanticRange's.